Update vaadin-archetype-application-example to use only Vaadin 8 API

ahie opened the pull request:
Closes vaadin/framework8-issues#348

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/archetypes/108)
<!-- Reviewable:end -->

denis-anisimov commented:
Reviewed 19 of 19 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks broke.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/authentication/LoginScreen.java, line 113 at r1](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KWqBdI6Z7hiwPnZbwFH:-KWqBdI6Z7hiwPnZbwFI:b-yhboqd) ([raw file](https://github.com/vaadin/archetypes/blob/aa8bd1953c21e8f1f8aa00b76fbfb04568b3d77b/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/authentication/LoginScreen.java#L113)):*
> ```Java
>                         + "Log in as &quot;admin&quot; to have full access. Log in with any other username to have read-only access. For all users, any password is fine",
>                 ContentMode.HTML);
>         loginInfoText.setSizeFull();
> ```

Looks like it's not necessary (at least just because of migration from V7 to V8).
What's the reason of the change ?

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/NumberField.java, line 14 at r1](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KWqCVbOxUgtOq-1Izpf:-KWqCVbPtUAowE6_J-Ve:b-hn6odi) ([raw file](https://github.com/vaadin/archetypes/blob/aa8bd1953c21e8f1f8aa00b76fbfb04568b3d77b/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/NumberField.java#L14)):*
> ```Java
>  * instead of the normal one.
>  */
> public class NumberField extends TextField {
> ```

This class has been removed from Spring CRUD demo.
Don't you think that it should be removed here as well.

The only reason why to keep it  is `AttributeExtension` (may be it's something that didn't exist in the version which I concert from original CRUD sample, or I just missed it).
But there is only one place where this field is used. So plain text field may be configured with the extension where it is needed in place instead of creating new component class just because of the extension. 

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductDataProvider.java, line 78 at r1](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KWqDrSb2XOEA-YA2je9:-KWqDrSb2XOEA-YA2jeA:bmkjs7y) ([raw file](https://github.com/vaadin/archetypes/blob/aa8bd1953c21e8f1f8aa00b76fbfb04568b3d77b/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductDataProvider.java#L78)):*
> ```Java
>     private boolean passesFilter(Object object) {
>         return object != null
>                 && object.toString().toLowerCase().contains(filterText);
> ```

`toLowerCase()` always requires locale. Use Locale.ENGLISH

Also there is an issue: one side (toString()) is always converted to lower case but the `filterText` doesn't have to be in lower case. It should be converted either here or in the setter.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java, line 36 at r1](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KWqFj-jA5TqvXTVHWBD:-KWqFj-jA5TqvXTVHWBE:b-yo1ic) ([raw file](https://github.com/vaadin/archetypes/blob/aa8bd1953c21e8f1f8aa00b76fbfb04568b3d77b/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java#L36)):*
> ```Java
>         decimalFormat.setMaximumFractionDigits(2);
>         decimalFormat.setMinimumFractionDigits(2);
>         addColumn("Price",
> ```

NumberRenderer is used in Spring CRUD.

This is again either recent update in the original sample app or something that I missed on converting.

Using this approach you should always keep in mind the ordering issue. 
When you use Renderer (or e.g. converter) then you just change the representation of the initial value. But initial value is still in place: you may filter/sort by the initial value not against its representation.
Here you are changing the value in the Grid so that it becomes the source of authority and there is no anymore access to initial value. So sort functionality which is given by the Grid automatically won't work (will work incorrectly).

But it does work (actually I don't know the current status of this feature) if you use real value with a Renderer.

So please either use NumberRenderer as it's done in Spring CRUD or extends this renderer for your purpose (might be tricky).

But actually this can be kept as is : I see there are methods in Spring CRUD which I wasn't aware of.
As I understand`Column.setSortProperty` should allow use any value for the column model but makes sorting works as expected.

Check for other columns that sorting works for them as well as expected. May be `setSortProperty` is required for any column to support ordering.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java, line 41 at r1](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KWqKJIORitzPAUg3J0F:-KWqKJIORitzPAUg3J0G:ba6hr6u) ([raw file](https://github.com/vaadin/archetypes/blob/aa8bd1953c21e8f1f8aa00b76fbfb04568b3d77b/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java#L41)):*
> ```Java
> 
>         // Add an traffic light icon in front of availability
>         addColumn("Availability", this::htmlFormatAvailability,
> ```

`setSortProperty`

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java, line 45 at r1](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KWqKQR7Sunv3592c4xQ:-KWqKQR7Sunv3592c4xR:bkkuem6) ([raw file](https://github.com/vaadin/archetypes/blob/aa8bd1953c21e8f1f8aa00b76fbfb04568b3d77b/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java#L45)):*
> ```Java
> 
>         // Show empty stock as "-"
>         addColumn("Stock Count", product -> {
> ```

`.setSortProperty`

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java, line 90 at r1](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KWqMSaghHH5qzookBaR:-KWqMSaghHH5qzookBaS:b5cgvpc) ([raw file](https://github.com/vaadin/archetypes/blob/aa8bd1953c21e8f1f8aa00b76fbfb04568b3d77b/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java#L90)):*
> ```Java
>     }
> 
>     public void deleteProduct(Product product) {
> ```

This method just shows a notification.
Is it needed at all now ?
There should be some other method which actually delete product from backend.

The name of the method is at least confusing. But I have strong feeling that this method is not needed at all (may be with some changes in workflow).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/archetypes/108)*
<!-- Sent from Reviewable.io -->

ahie commented:
Review status: 11 of 20 files reviewed at latest revision, 7 unresolved discussions.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/authentication/LoginScreen.java, line 113 at r1](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KWqBdI6Z7hiwPnZbwFH:-KWrY2pQL4_IkxNI5exC:b-6sg4aw) ([raw file](https://github.com/vaadin/archetypes/blob/aa8bd1953c21e8f1f8aa00b76fbfb04568b3d77b/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/authentication/LoginScreen.java#L113)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary>
> Looks like it's not necessary (at least just because of migration from V7 to V8).
> What's the reason of the change ?
</details>
Label's constructor no longer sets the width to full by default in v8.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/NumberField.java, line 14 at r1](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KWqCVbOxUgtOq-1Izpf:-KWrYk1pVCPeWrAOL65A:b-9znjar) ([raw file](https://github.com/vaadin/archetypes/blob/aa8bd1953c21e8f1f8aa00b76fbfb04568b3d77b/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/NumberField.java#L14)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary>
> This class has been removed from Spring CRUD demo.
> Don't you think that it should be removed here as well.
> 
> The only reason why to keep it  is `AttributeExtension` (may be it's something that didn't exist in the version which I concert from original CRUD sample, or I just missed it).
> But there is only one place where this field is used. So plain text field may be configured with the extension where it is needed in place instead of creating new component class just because of the extension. 
</details>
Removed. AttributeExtension now set in the ProductForm.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductDataProvider.java, line 78 at r1](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KWqDrSb2XOEA-YA2je9:-KWrZ3Xl6eJptexiN4lH:b-896fix) ([raw file](https://github.com/vaadin/archetypes/blob/aa8bd1953c21e8f1f8aa00b76fbfb04568b3d77b/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductDataProvider.java#L78)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary>
> `toLowerCase()` always requires locale. Use Locale.ENGLISH
> 
> Also there is an issue: one side (toString()) is always converted to lower case but the `filterText` doesn't have to be in lower case. It should be converted either here or in the setter.
</details>
Done.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java, line 36 at r1](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KWqFj-jA5TqvXTVHWBD:-KX5u_WhFzbsfNZt-c6F:b30kwtb) ([raw file](https://github.com/vaadin/archetypes/blob/aa8bd1953c21e8f1f8aa00b76fbfb04568b3d77b/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java#L36)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary>
> NumberRenderer is used in Spring CRUD.
> 
> This is again either recent update in the original sample app or something that I missed on converting.
> 
> Using this approach you should always keep in mind the ordering issue. 
> When you use Renderer (or e.g. converter) then you just change the representation of the initial value. But initial value is still in place: you may filter/sort by the initial value not against its representation.
> Here you are changing the value in the Grid so that it becomes the source of authority and there is no anymore access to initial value. So sort functionality which is given by the Grid automatically won't work (will work incorrectly).
> 
> But it does work (actually I don't know the current status of this feature) if you use real value with a Renderer.
> 
> So please either use NumberRenderer as it's done in Spring CRUD or extends this renderer for your purpose (might be tricky).
> 
> But actually this can be kept as is : I see there are methods in Spring CRUD which I wasn't aware of.
> As I understand`Column.setSortProperty` should allow use any value for the column model but makes sorting works as expected.
> 
> Check for other columns that sorting works for them as well as expected. May be `setSortProperty` is required for any column to support ordering.
</details>
I opted to use setComparator for columns. Differs from the spring demo, since backend sorting can be easily done with property ids there.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java, line 41 at r1](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KWqKJIORitzPAUg3J0F:-KWrb2b0aQJvng8cWRz1:bjdc2o9) ([raw file](https://github.com/vaadin/archetypes/blob/aa8bd1953c21e8f1f8aa00b76fbfb04568b3d77b/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java#L41)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary>
> `setSortProperty`
</details>
Done, as setComparator.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java, line 45 at r1](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KWqKQR7Sunv3592c4xQ:-KWrb3hoHYIypJYNzZlI:bjdc2o9) ([raw file](https://github.com/vaadin/archetypes/blob/aa8bd1953c21e8f1f8aa00b76fbfb04568b3d77b/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java#L45)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary>
> `.setSortProperty`
</details>
Done, as setComparator.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java, line 90 at r1](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KWqMSaghHH5qzookBaR:-KX5TTvniUr_y5byKAxE:b-u9wxot) ([raw file](https://github.com/vaadin/archetypes/blob/aa8bd1953c21e8f1f8aa00b76fbfb04568b3d77b/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java#L90)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary>
> This method just shows a notification.
> Is it needed at all now ?
> There should be some other method which actually delete product from backend.
> 
> The name of the method is at least confusing. But I have strong feeling that this method is not needed at all (may be with some changes in workflow).
</details>
This method contains a call to view.removeProduct which in turn calls the data provider's delete method, same as in the spring demo.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/archetypes/108)*
<!-- Sent from Reviewable.io -->

ahie commented:
Review status: 11 of 20 files reviewed at latest revision, 9 unresolved discussions.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductForm.java, line 121 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX6AT6m7j_BSERLKDeQ:-KX6AT6m7j_BSERLKDeR:b-ayq8el) ([raw file](https://github.com/vaadin/archetypes/blob/e42aecb2bf6c82afc354620e8df57fab06f26b57/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductForm.java#L121)):*
> ```Java
>         }
>         currentProduct = product;
>         beanBinder.readBean(product);
> ```

Tested with https://dev.vaadin.com/review/#/c/14792/ applied to local snapshot.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudView.java, line 136 at r1](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX6AZmagn1sRFbbOZG4:-KX6AZmagn1sRFbbOZG5:b-66xu8a) ([raw file](https://github.com/vaadin/archetypes/blob/aa8bd1953c21e8f1f8aa00b76fbfb04568b3d77b/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudView.java#L136)):*
> ```Java
>         } else {
>             form.removeStyleName("visible");
>             // FIXME: Issue #286
> ```

Tested with https://dev.vaadin.com/review/#/c/14454/ applied to local snapshot.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/archetypes/108)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 8 of 9 files at r2.
Review status: 19 of 20 files reviewed at latest revision, 12 unresolved discussions.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductForm.java, line 72 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX9q0ulYfUeLd7R17ei:-KX9q0ulYfUeLd7R17ej:brop7t) ([raw file](https://github.com/vaadin/archetypes/blob/e42aecb2bf6c82afc354620e8df57fab06f26b57/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductForm.java#L72)):*
> ```Java
>         // Mark the stock count field as numeric.
>         // This affects the virtual keyboard shown on mobile devices.
>         AttributeExtension ae = new AttributeExtension();
> ```

Bad variable name.
Minor.
But please fix it if you are going to make other changes......

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductForm.java, line 90 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX9m0v1uEu-wsnxhnRr:-KX9m0v1uEu-wsnxhnRs:bfc7hlg) ([raw file](https://github.com/vaadin/archetypes/blob/e42aecb2bf6c82afc354620e8df57fab06f26b57/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductForm.java#L90)):*
> ```Java
> 
>         // enable/disable save button while editing
>         beanBinder.addStatusChangeListener(event -> {
> ```

What about other buttons ?
Delete/cancel/revert .....

See Spring CRUD.
At least one of those buttons should also be updated when status is changed.
(Perhaps I've introduced new "revert" button to demonstrate this : it should be enabled if there are any changes regardless whether they are valid or not. And this might not be possible with https://dev.vaadin.com/review/#/c/14792/  patch set. See my comments there).

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductForm.java, line 96 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX9lsJXcNIAo4ybeoxE:-KX9lsJXcNIAo4ybeoxF:bkwmhqp) ([raw file](https://github.com/vaadin/archetypes/blob/e42aecb2bf6c82afc354620e8df57fab06f26b57/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductForm.java#L96)):*
> ```Java
>         });
> 
>         save.addClickListener(click -> {
> ```

`click` is misleading here. It's an event. Though it's not used anyway.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java, line 41 at r1](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KWqKJIORitzPAUg3J0F:-KX9oX8lDEd7Zf2iZ1TI:bcwcl62) ([raw file](https://github.com/vaadin/archetypes/blob/aa8bd1953c21e8f1f8aa00b76fbfb04568b3d77b/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java#L41)):*
<details><summary><i>Previously, ahie (Aleksi Hietanen) wrote…</i></summary>
> Done, as setComparator.
</details>
Same question here: why not `setSortProperty` ?

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java, line 45 at r1](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KWqKQR7Sunv3592c4xQ:-KX9nvZgA0ysManOKGP9:b-ckjboc) ([raw file](https://github.com/vaadin/archetypes/blob/aa8bd1953c21e8f1f8aa00b76fbfb04568b3d77b/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java#L45)):*
<details><summary><i>Previously, ahie (Aleksi Hietanen) wrote…</i></summary>
> Done, as setComparator.
</details>
Again : why not `setSortProperty`  ?

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java, line 38 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX9nH2tMlMM8fcwWUc5:-KX9nH2tMlMM8fcwWUc6:b-tq0i5w) ([raw file](https://github.com/vaadin/archetypes/blob/e42aecb2bf6c82afc354620e8df57fab06f26b57/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java#L38)):*
> ```Java
>         addColumn("Price",
>                 product -> decimalFormat.format(product.getPrice()) + " €")
>                         .setComparator((p1, p2) -> {
> ```

Do you really need to use comparator here ?
Doesn't `setSortProperty` work ?

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java, line 45 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX9nfFCmC5IRh-u9LlC:-KX9nfFCmC5IRh-u9LlD:b-6n5l4v) ([raw file](https://github.com/vaadin/archetypes/blob/e42aecb2bf6c82afc354620e8df57fab06f26b57/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java#L45)):*
> ```Java
>         addColumn("Availability", this::htmlFormatAvailability,
>                 new HtmlRenderer()).setComparator((p1, p2) -> {
>                     return p1.getAvailability().toString()
> ```

Why `toString()` ? enumerations are comparable.
And I think you can use `Comparator. naturalOrder()` almost everywhere.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java, line 76 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX9o9ROHe0lUZKohScr:-KX9o9ROHe0lUZKohScs:by3j5as) ([raw file](https://github.com/vaadin/archetypes/blob/e42aecb2bf6c82afc354620e8df57fab06f26b57/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java#L76)):*
> ```Java
> 
>         String color = "";
>         if (availability == Availability.AVAILABLE) {
> ```

Let's use switch here.
More readable.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java, line 90 at r1](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KWqMSaghHH5qzookBaR:-KX9oix-8QU6XQiJK5Y3:bwa97x3) ([raw file](https://github.com/vaadin/archetypes/blob/aa8bd1953c21e8f1f8aa00b76fbfb04568b3d77b/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java#L90)):*
<details><summary><i>Previously, ahie (Aleksi Hietanen) wrote…</i></summary>
> This method contains a call to view.removeProduct which in turn calls the data provider's delete method, same as in the spring demo.
</details>
Alright......
My bad. I haven't noticed the code below the changed line ( reviewable didn't show me it). So I decided that this is one line method.
OK

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java, line 43 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX9qjQrfx-ZuZk4U6tT:-KX9qjQsfVq8_q2f2_KM:beedld4) ([raw file](https://github.com/vaadin/archetypes/blob/48849aaba2670940ce54fd56e0d41e611bc85a88/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java#L43)):*
> ```Java
>         setFragmentParameter("");
>         view.clearSelection();
>         view.editProduct(null);
> ```

A question : what causes this logic in the current implementation ? (cancel editing, equivalent of editProduct(null) ?

And is there any reason to accept `null` for `editProduct` method ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/archetypes/108)*
<!-- Sent from Reviewable.io -->

ahie commented:
Review status: 15 of 20 files reviewed at latest revision, 11 unresolved discussions.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductForm.java, line 72 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX9q0ulYfUeLd7R17ei:-KXAxt6wsRjbfkkG2j2S:b-896fix) ([raw file](https://github.com/vaadin/archetypes/blob/e42aecb2bf6c82afc354620e8df57fab06f26b57/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductForm.java#L72)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary>
> Bad variable name.
> Minor.
> But please fix it if you are going to make other changes......
</details>
Done.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductForm.java, line 90 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX9m0v1uEu-wsnxhnRr:-KXA-2G-_da0VgjHd_bh:b-oixg4) ([raw file](https://github.com/vaadin/archetypes/blob/e42aecb2bf6c82afc354620e8df57fab06f26b57/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductForm.java#L90)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary>
> What about other buttons ?
> Delete/cancel/revert .....
> 
> See Spring CRUD.
> At least one of those buttons should also be updated when status is changed.
> (Perhaps I've introduced new "revert" button to demonstrate this : it should be enabled if there are any changes regardless whether they are valid or not. And this might not be possible with https://dev.vaadin.com/review/#/c/14792/  patch set. See my comments there).
</details>
Delete should always be enabled in my opinion. Since cancel is only used for closing the form it should also be always enabled. Added a discard button like in the spring version.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductForm.java, line 96 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX9lsJXcNIAo4ybeoxE:-KXAyTyae6xMqgfDG24u:b-896fix) ([raw file](https://github.com/vaadin/archetypes/blob/e42aecb2bf6c82afc354620e8df57fab06f26b57/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductForm.java#L96)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary>
> `click` is misleading here. It's an event. Though it's not used anyway.
</details>
Done.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java, line 38 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX9nH2tMlMM8fcwWUc5:-KXA1nd0oJu1lqvukzvW:b-sq4jt5) ([raw file](https://github.com/vaadin/archetypes/blob/e42aecb2bf6c82afc354620e8df57fab06f26b57/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java#L38)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary>
> Do you really need to use comparator here ?
> Doesn't `setSortProperty` work ?
</details>
`setSortProperty` effectively propagates a string to the data provider via the query object, so by using `setComparator` we don't have to deal with a string in the data provider (which works well with spring, but not with the custom "backend" of this archetype)

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java, line 45 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX9nfFCmC5IRh-u9LlC:-KXA-oq2t_OmDCd_Ovuq:b-p1mitm) ([raw file](https://github.com/vaadin/archetypes/blob/e42aecb2bf6c82afc354620e8df57fab06f26b57/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java#L45)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary>
> Why `toString()` ? enumerations are comparable.
> And I think you can use `Comparator. naturalOrder()` almost everywhere.
</details>
Results in the wrong ordering if not comparing the strings. Not sure how `Comparator.naturalOder()` could be used since `setComparator` only knows about the grid's type.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java, line 76 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX9o9ROHe0lUZKohScr:-KXA05dTig6OyOwkVSba:b-896fix) ([raw file](https://github.com/vaadin/archetypes/blob/e42aecb2bf6c82afc354620e8df57fab06f26b57/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java#L76)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary>
> Let's use switch here.
> More readable.
</details>
Done.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java, line 43 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX9qjQrfx-ZuZk4U6tT:-KXA0CltHbRg3THrvqAf:b9jors8) ([raw file](https://github.com/vaadin/archetypes/blob/48849aaba2670940ce54fd56e0d41e611bc85a88/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java#L43)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary>
> A question : what causes this logic in the current implementation ? (cancel editing, equivalent of editProduct(null) ?
> 
> And is there any reason to accept `null` for `editProduct` method ?
</details>
Cancels editing and editProduct(null) is how this has been done in the logic previously.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/archetypes/108)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 2 of 4 files at r3.
Review status: 14 of 20 files reviewed at latest revision, 5 unresolved discussions.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductForm.java, line 96 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX9lsJXcNIAo4ybeoxE:-KXB1vowI6INrBn7Dihc:b-ui4e8w) ([raw file](https://github.com/vaadin/archetypes/blob/e42aecb2bf6c82afc354620e8df57fab06f26b57/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductForm.java#L96)):*
<details><summary><i>Previously, ahie (Aleksi Hietanen) wrote…</i></summary>
> Done.
</details>
Not done, but minor.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java, line 45 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX9nfFCmC5IRh-u9LlC:-KXB2MaJsq5AkC0X69fR:b-x4ntz8) ([raw file](https://github.com/vaadin/archetypes/blob/e42aecb2bf6c82afc354620e8df57fab06f26b57/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductGrid.java#L45)):*
<details><summary><i>Previously, ahie (Aleksi Hietanen) wrote…</i></summary>
> Results in the wrong ordering if not comparing the strings. Not sure how `Comparator.naturalOder()` could be used since `setComparator` only knows about the grid's type.
</details>
OK.... toString() is implemented in a special way in `Availability`. OK

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java, line 43 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX9qjQrfx-ZuZk4U6tT:-KXB2nLVjIFcjqkvzo8o:b-yy5onj) ([raw file](https://github.com/vaadin/archetypes/blob/48849aaba2670940ce54fd56e0d41e611bc85a88/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java#L43)):*
<details><summary><i>Previously, ahie (Aleksi Hietanen) wrote…</i></summary>
> Cancels editing and editProduct(null) is how this has been done in the logic previously.
</details>
The question was: what's the replacement for this currently ?
Why there is no need in `editProduct(null)` anymore ?
And is it still necessary to accept `null` in `editProduct` method now if this invocation is removed ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/archetypes/108)*
<!-- Sent from Reviewable.io -->

ahie commented:
Review status: 14 of 20 files reviewed at latest revision, 1 unresolved discussion.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductForm.java, line 96 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX9lsJXcNIAo4ybeoxE:-KXB3w5bOQVfxCOicUsO:b-5zwkyx) ([raw file](https://github.com/vaadin/archetypes/blob/e42aecb2bf6c82afc354620e8df57fab06f26b57/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/ProductForm.java#L96)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary>
> Not done, but minor.
</details>
Oops, forgot to add those changes. Done now.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java, line 43 at r2](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KX9qjQrfx-ZuZk4U6tT:-KXB3HDpzIFmbDVuCFOh:b-eptnlb) ([raw file](https://github.com/vaadin/archetypes/blob/48849aaba2670940ce54fd56e0d41e611bc85a88/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java#L43)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary>
> The question was: what's the replacement for this currently ?
> Why there is no need in `editProduct(null)` anymore ?
> And is it still necessary to accept `null` in `editProduct` method now if this invocation is removed ?
</details>
Oh, that is eventually called by the selection change listener.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/archetypes/108)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 1 of 3 files at r4.
Review status: 15 of 20 files reviewed at latest revision, 2 unresolved discussions.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java, line 85 at r4](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KXB8MLfCxl3RtPWIdUn:-KXB8MLfCxl3RtPWIdUo:b58ypt1) ([raw file](https://github.com/vaadin/archetypes/blob/7aa01c1db8f3d9e8f5456093bd9ad5d4888f7b34/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java#L85)):*
> ```Java
>                 + product.getId() + ") updated");
>         view.clearSelection();
>         view.editProduct(null);
> ```

In addition to above: then this is not needed as well. 
`clearSelection` should imply this, right ?

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java, line 95 at r4](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KXB8YbaoHXZtdUQ6GGH:-KXB8YbaoHXZtdUQ6GGI:b83vwvw) ([raw file](https://github.com/vaadin/archetypes/blob/7aa01c1db8f3d9e8f5456093bd9ad5d4888f7b34/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java#L95)):*
> ```Java
> 
>         view.clearSelection();
>         view.editProduct(null);
> ```

And here:
In addition to above: then this is not needed as well. 
`clearSelection` should imply this, right ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/archetypes/108)*
<!-- Sent from Reviewable.io -->

ahie commented:
Review status: 15 of 20 files reviewed at latest revision, 2 unresolved discussions.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java, line 85 at r4](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KXB8MLfCxl3RtPWIdUn:-KXBb1b34eESHEgESiFp:b-896fix) ([raw file](https://github.com/vaadin/archetypes/blob/7aa01c1db8f3d9e8f5456093bd9ad5d4888f7b34/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java#L85)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary>
> In addition to above: then this is not needed as well. 
> `clearSelection` should imply this, right ?
</details>
Done.

---

*[vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java, line 95 at r4](https://reviewable.io:443/reviews/vaadin/archetypes/108#-KXB8YbaoHXZtdUQ6GGH:-KXBb23Zd5D5vVxDMUJ2:b-896fix) ([raw file](https://github.com/vaadin/archetypes/blob/7aa01c1db8f3d9e8f5456093bd9ad5d4888f7b34/vaadin-archetype-application-example/src/main/resources/archetype-resources/__rootArtifactId__-ui/src/main/java/samples/crud/SampleCrudLogic.java#L95)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary>
> And here:
> In addition to above: then this is not needed as well. 
> `clearSelection` should imply this, right ?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/archetypes/108)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 1 of 1 files at r5.
Review status: 15 of 20 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/archetypes/108)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 1 of 9 files at r2, 2 of 4 files at r3, 2 of 3 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/archetypes/108)*
<!-- Sent from Reviewable.io -->
